thunder/core/module.py

```diff
@@ -141,6 +143,117 @@ def load_original_state_dict(self, state_dict):
                 else:
                     raise NotImplementedError(f"don't know how to handle {full_k}")
 
+    def state_dict(self, *, destination=None, prefix="", keep_vars=False):
```

Comment:
at glance, the major difference from `torch.nn.Module` seems to be that this implementation does not use `_state_dict_pre_hooks` and `_state_dict_hooks`.
I thought either of them could be used for `thunder.distributed.fsdp` to register its hook to create a state dict of unsharded params.

Reply:
So the tricky thing is that in order to support the hooks, we would need to go to a module-by-module thing again and probably create a copy of the module with the overrides as parameters.
This would be a more elaborate than the current version, but should be doable.
We needed a thing that can save / load state dicts rather soon, so I suggest to start with this and refine from there.

But I wonder: What this does (and what we currently need) is a way to get the state of the thunder module. With FSDP, I would argue that that is the sharded state, and that we should make a function or method to get the unsharded state. Also, in my imagination, such a thing would do communication and I wonder if that is conceptually heavy-handed to do in hooks.

Reply:
> We needed a thing that can save / load state dicts rather soon, so I suggest to start with this and refine from there.

agreed.

> With FSDP, I would argue that that is the sharded state, and that we should make a function or method to get the unsharded state.

yes.

> I wonder if that is conceptually heavy-handed to do in hooks.

I embarrassingly am not aware of any criteria to decide if a workload is fine to enclosed in a hook or not so I thought it's fair.

Reply:
It's likely a bit of a taste thing and the thresholds differ for people, too.

But so in some sense hooks are hidden code, so to my mind, the answer is somewhere between "the hooks should not be used by library authors at all, only by end users" (I think I wrote something like that on the PyTorch forums a long time ago) and the hooks, as hidden code, should have only niche effects.

I think for loading state dicts, the hooks/customization is there (and used) in particular to facilitate loading state dicts from older versions.